crates/datasources/src/sqlserver/mod.rs

```diff
@@ -0,0 +1,559 @@
+        limit: Option<usize>,
+    ) -> DatafusionResult<Arc<dyn ExecutionPlan>> {
+        // Project the schema.
+        let projected_schema = match projection {
+            Some(projection) => Arc::new(self.arrow_schema.project(projection)?),
+            None => self.arrow_schema.clone(),
+        };
+
+        // Get the projected columns, joined by a ','. This will be put in the
+        // 'SELECT ...' portion of the query.
+        let projection_string = projected_schema
+            .fields
+            .iter()
+            .map(|f| f.name().clone())
+            .collect::<Vec<_>>()
+            .join(",");
+
+        let limit_string = match limit {
+            Some(limit) => format!("LIMIT {}", limit),
+            None => String::new(),
+        };
+
+        // TODO: Where/filters
+
+        let query = format!(
+            "SELECT {projection_string} FROM {}.{} {limit_string}",
+            self.schema, self.table
+        );
+
+        Ok(Arc::new(SqlServerExec {
+            query,
+            state: self.state.clone(),
+            arrow_schema: projected_schema,
+            metrics: ExecutionPlanMetricsSet::new(),
+        }))
+    }
+
+    async fn insert_into(
+        &self,
+        _state: &SessionState,
+        _input: Arc<dyn ExecutionPlan>,
+        _overwrite: bool,
+    ) -> DatafusionResult<Arc<dyn ExecutionPlan>> {
+        Err(DataFusionError::Execution(
+            "inserts not supported for SQL Server".to_string(),
+        ))
+    }
+}
+
+/// Execution plan for reading from SQL Server.
+struct SqlServerExec {
+    query: String,
+    state: Arc<SqlServerAccessState>,
+    arrow_schema: ArrowSchemaRef,
+    metrics: ExecutionPlanMetricsSet,
+}
+
+impl ExecutionPlan for SqlServerExec {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> ArrowSchemaRef {
+        self.arrow_schema.clone()
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        Partitioning::UnknownPartitioning(1)
+    }
+
+    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
+        None
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        Vec::new()
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        _children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> DatafusionResult<Arc<dyn ExecutionPlan>> {
+        Err(DataFusionError::Execution(
+            "cannot replace children for SqlServerExec".to_string(),
+        ))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> DatafusionResult<SendableRecordBatchStream> {
+        if partition != 0 {
+            return Err(DataFusionError::Execution(
+                "only single partition supported".to_string(),
+            ));
+        }
+
+        // Clones to ensure the future is static.
+        let query = self.query.clone();
+        let state = self.state.clone();
+        let fut = async move { state.client.query(query).await };
```

Comment:
Only 331 could be inline, the others would need to stay here. I wanted to keep these grouped together to have the comment there to avoid someone coming in a making future not static and wondering where all the errors are coming from.